contracts/Marketplace.sol

```diff
@@ -373,24 +376,43 @@ contract Marketplace is Proofs, StateRetrieval, Endian {
     address withdrawRecipient
   ) public {
     Request storage request = _requests[requestId];
-    require(
-      block.timestamp > requestExpiry(requestId),
-      "Request not yet timed out"
-    );
     require(request.client == msg.sender, "Invalid client address");
     RequestContext storage context = _requestContexts[requestId];
-    require(context.state == RequestState.New, "Invalid state");
+    RequestState state = requestState(requestId);
+    require(
+      state == RequestState.Cancelled ||
+        state == RequestState.Failed ||
+        state == RequestState.Finished,
+      "Invalid state"
+    );
+    require(context.fundsToReturnToClient != 0, "Nothing to withdraw");
+
+    if (state == RequestState.Cancelled) {
+      context.state = RequestState.Cancelled;
+      emit RequestCancelled(requestId);
+
+      // `fundsToReturnToClient` currently tracks funds to be returned for requests that successfully finish.
+      // When requests are cancelled, funds earmarked for payment for the duration
+      // between request expiry and request end (for every slot that was filled), should be returned to the client.
+      // Update `fundsToReturnToClient` to reflect this.
+      context.fundsToReturnToClient +=
```

Comment:
> Doesn't this mean that the amount of funds that a client receives depends on when it calls withdrawFunds().

No, it does not depend on that. You are most probably referencing the `context.slotsFilled` right? That variable is not lowered when calling `freeSlot` in states **Failed/Cancelled/Finished**, only when the node would call that function during "running" of the Request, which again would not effect this function as it can be called only in those states...

Reply:
No, I didn't mean `context.slotsFilled`. I was thinking about the additions to `context.fundsToReturnToClient` in `freeSlot`. But looking through the code again, I realize that this only happens in `_forciblyFreeSlot()`, and that can only happen while the slot is Filled.